### PR TITLE
Features/sdk endpoint to retreive contract info

### PR DESCRIPTION
**Pull request by aneyzberg**
Problem
=======
We need to be able to retrieve a contract addresses based on a contract's name. 
We also need to be able to run integration tests to prove test that we can retrieve a contract address based on a contract name from the chain
[link to Pivotal Tracker #177584391](https://www.pivotaltracker.com/story/show/177584391)
[link to Pivotal Tracker #177144711](https://www.pivotaltracker.com/story/show/177144711)

Solution
========
- Add functionality to retrieve contract address from chain.
- add integration testing of contract.
   - We added snapshot testing to be able to replay chain state so we could use the same process for many tests
   - We added a docker image we pushed up to docker hub called `unfinishedlabs/ganache-truffle-node:latest`  - that spins up a node so we can deploy contracts to it in order to run tests against our contracts.
- update github actions config to set up integration testing.

with @acruikshank 

Change summary:
---------------
* Add a `getContractAddress` function in `contract module` that will retrieve contract address from chain based on name of contract
* Add snapshot testing for `contract` module so we can replay state chain state in our tests
* Update `main.yml` to setup integration testing for sdk
* Create docker image called `unfinishedlabs/ganache-truffle-node:latest` that is pushed to docker hub so we can run an truffle node to deploy our contracts to during CI
* add typechain  to repo so we can bring in types from Contracts repo

### CODE REVIEW

File: src/contract/contract.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+export const keccakTopic = (topic: string) => "0x" + keccak256(topic);
+
+interface ContractResult {
+  contractAddress: string;
+  contractName: string;
+  blockNumber: number;
+  blockHash: string;
+}
+
+export const DSNPMigrationABI: AbiInput[] = [
```
Comment (wilwade) [inline src/contract/contract.ts:17]: This can also now be imported from `@unfinishedlabs/contracts`

File: src/contract/contract.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  contractName: string
+): Promise<ContractResult | Record<string, string>> => {
+  const topic = keccakTopic("DSNPMigration(address,string)");
+
+  const logs: Log[] = await web3.eth.getPastLogs({ topics: [topic], fromBlock: 0 });
+  const decodedValues = decodeReturnValues(web3, DSNPMigrationABI, topic, logs);
+  const filteredResults: ContractResult[] = filterValues(decodedValues, contractName);
+  return filteredResults.length > 0
+    ? filteredResults[filteredResults.length - 1]
+    : { error: `No longs found for contract name: ${contractName}` };
```
Comment (wilwade) [inline src/contract/contract.ts:61]: This is an odd way to pass the error back for ts. I think I would have expected either a `null` or a `throw`

**Review by wilwade — COMMENTED** [review]
Two quick notes just on the contract.ts side of things.

File: src/contract/contract.ts (hunk 2 above)
Comment (aneyzberg) [reply src/contract/contract.ts:61]: and that is true. this should not actually be an error. Just return an empty string/

File: src/contract/contract.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  },
+  {
+    indexed: false,
+    internalType: "string",
+    name: "contractName",
+    type: "string",
+  },
+];
+
+const decodeReturnValues = (web3: Web3, inputs: AbiInput[], topic: string, logs: Log[]) => {
```
Comment (enddynayn) [inline src/contract/contract.ts:32]: perhaps we can use "named parameters"

File: src/contract/contract.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  return logs.map((log: Log) => {
+    const decodedValue = web3.eth.abi.decodeLog(inputs, log.data, [topic]);
```
Comment (enddynayn) [inline src/contract/contract.ts:34]: alternative way

```
const { contractAddress, contractName, } = web3.eth.abi.decodeLog(inputs, log.data, [topic]);
const { blockNumber, blockHash } = log;

 return {
      contractAddress,
      contractName,
      blockNumber,
      blockHash
    };
```

File: src/contracts/announcement.ts [hunk 5]
```diff
@@ -7,54 +7,17 @@ import { HexString } from "../types/Strings";
...
+import { abi as announcerABI } from "@unfinishedlabs/contracts/abi/Announcer.json";
 
 const GAS_LIMIT_BUFFER = 1000;
 const contract: Contract | null = null;
 
-const getContract = (web3Instance: Web3): Contract => {
+const getContract = async (web3Instance: Web3): Promise<any> => {
   if (contract) return contract;
-  return new web3Instance.eth.Contract(BATCH_CONTRACT_ABI, BATCH_CONTRACT_ADDRESS);
+  const contractInfo = (await getContractAddress(web3Instance, "Announcer")) as ContractResult;
```
Comment (enddynayn) [inline src/contracts/announcement.ts:19]: I am new to typescript. So the following is a question. Can you do?

```
const contractInfo: ContractResult = await getContractAddress(web3Instance, "Announcer")
```

**Review by enddynayn — APPROVED** [review]

File: src/contract/contract.ts (hunk 3 above)
Comment (aneyzberg) [reply src/contract/contract.ts:32]: we had gone back and forth about this a while ago. and I think the decision was to be explicit for now?

File: package.json [hunk 6]
```diff
@@ -90,6 +94,9 @@
   },
   "dependencies": {
     "js-sha3": "^0.8.0",
-    "web3": "^1.3.5"
+    "node-fetch": "^2.6.1",
```
Comment (wilwade) [inline package.json:97]: Do we use this anywhere?

**Review by wilwade — DISMISSED** [review]
Just one question about node-fetch.

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -25,19 +35,57 @@ jobs:
...
+        run: |
+          echo `npm list @unfinishedlabs/contracts --verbose`
+          export version=`npm list @unfinishedlabs/contracts --silent | sed -En 's/.*contracts@0.0.0-(\w+)\s.*$/\1/p'`
+          export result=`git ls-remote https://github.com/LibertyDSNP/contracts.git | grep -m 1 $version | cut -c 1-40`
+          echo "contract_version=$result" >> $GITHUB_ENV
+        shell: bash
+        working-directory: ./main
+
+      - name: Check result
+        run:  echo "${{ env.contract_version }}"
```
Comment (rosalinekarr) [inline .github/workflows/main.yml:51]: Should we take this step out? Or, if we leave it in, could we make the message more descriptive, like "Using contract version: ${{ env.contract_version }}"?

File: src/contracts/announcement.ts [hunk 8]
```diff
@@ -1,63 +1,28 @@
...
+const contract: Announcer | null = null;
+const CONTRACT_NAME = "Announcer";
 
-const getContract = (web3Instance: Web3): Contract => {
+const getContract = async (web3Instance: Web3): Promise<Announcer | null> => {
   if (contract) return contract;
-  return new web3Instance.eth.Contract(BATCH_CONTRACT_ABI, BATCH_CONTRACT_ADDRESS);
+  const contractAddr = await getContractAddress(web3Instance, CONTRACT_NAME);
+  if (!contractAddr) return null;
+  return (new web3Instance.eth.Contract(announcerABI as AbiItem[], contractAddr) as any) as Announcer;
```
Comment (rosalinekarr) [inline src/contracts/announcement.ts:22]: `) as any) as Announcer` => `) as unknown) as Announcer`

File: src/contracts/announcement.ts [hunk 9]
```diff
@@ -1,63 +1,28 @@
...
-const contract: Contract | null = null;
+const contract: Announcer | null = null;
+const CONTRACT_NAME = "Announcer";
 
-const getContract = (web3Instance: Web3): Contract => {
+const getContract = async (web3Instance: Web3): Promise<Announcer | null> => {
   if (contract) return contract;
-  return new web3Instance.eth.Contract(BATCH_CONTRACT_ABI, BATCH_CONTRACT_ADDRESS);
+  const contractAddr = await getContractAddress(web3Instance, CONTRACT_NAME);
+  if (!contractAddr) return null;
```
Comment (rosalinekarr) [inline src/contracts/announcement.ts:21]: If we threw an error here, we could simplify the return type of the function to just `Promise<Announcer>`. No strong opinions though.

File: src/test/hardhatRPC.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+/* eslint @typescript-eslint/no-var-requires: "off" */
+require("dotenv").config();
+import Web3 from "web3";
+
+let latestSnapshot: string;
+
+export const snapshotHardhat = async () => {
```
Comment (rosalinekarr) [inline src/test/hardhatRPC.ts:7]: `async (): Promise<void> => {`

File: src/test/hardhatRPC.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  const provider = new Web3.providers.HttpProvider(process.env.RPC_URL as string);
+  const snapshotResponse: any = await new Promise((resolve, reject) =>
```
Comment (rosalinekarr) [inline src/test/hardhatRPC.ts:9]: Substitute `any` with `unknown`. [It's preferred in TypeScript now.](https://devblogs.microsoft.com/typescript/announcing-typescript-3-0-rc-2/#the-unknown-type)

File: src/test/hardhatRPC.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    provider.send({ jsonrpc: "2.0", method: "evm_snapshot", id: 1, params: [] }, (err, snapshotResponse) =>
+      err ? reject(err) : resolve(snapshotResponse)
+    )
+  );
+
+  expect(snapshotResponse.error).toBeUndefined();
+  latestSnapshot = snapshotResponse.result as string;
+};
+
+export const revertHardhat = async () => {
```
Comment (rosalinekarr) [inline src/test/hardhatRPC.ts:19]: `async (): Promise<void> => {`

File: src/test/hardhatRPC.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  const provider = new Web3.providers.HttpProvider(process.env.RPC_URL as string);
+  const revertResponse: any = await new Promise((resolve, reject) =>
```
Comment (rosalinekarr) [inline src/test/hardhatRPC.ts:21]: `any` => `unknown`

**Review by rosalinekarr — CHANGES_REQUESTED** [review]
A few small things, but otherwise lgtm

File: package.json (hunk 6 above)
Comment (aneyzberg) [reply package.json:97]: oh we did, but I don't think we do anymore. Will remove

File: .github/workflows/main.yml (hunk 7 above)
Comment (aneyzberg) [reply .github/workflows/main.yml:51]: yes my mistake. Meant to remove.

**Review by rosalinekarr — APPROVED** [review]

**Pull request merged**